### PR TITLE
Create CAPI releases

### PR DESCRIPTION
**Pull request by njuettner**
Towards: https://github.com/giantswarm/giantswarm/issues/31641

This addresses the creation of CAPI releases.

## A couple of changes:

- Dropped apps and components which aren't used anymore
- Added new apps and components which are now required for all CAPI providers
- Aligning the README.md structure with our current approach, example below
- Added the announcement.md
- Support AWS (CAPA), Azure (CAPZ), vSphere (CAPV) and cloud-director (CAPVCD) and dropped old vintage providers 

## Example

Using "bump all" to go through all apps and components to update them but use dedicated `--component` or `--app` to use specific user overwrite e.g. K8s 1.30.7

```
devctl release create --base v29.4.0 --name v29.5.0 --provider aws --bumpall --component kubernetes@1.30.7
```

Lists of changes:

```
+------------------------------------+---------------------+---------------------+
| APP NAME                           | CURRENT APP VERSION | DESIRED APP VERSION |
+------------------------------------+---------------------+---------------------+
| aws-ebs-csi-driver                 | 2.30.1              | Unchanged           |
| aws-ebs-csi-driver-servicemonitors | 0.1.0               | Unchanged           |
| aws-pod-identity-webhook           | 1.17.0              | Unchanged           |
| capi-node-labeler                  | 0.5.0               | Unchanged           |
| cert-exporter                      | 2.9.3               | Unchanged           |
| cert-manager                       | 3.8.1               | 3.8.2               |
| chart-operator-extensions          | 1.1.2               | Unchanged           |
| cilium                             | 0.25.1              | 0.29.0              |
| cilium-crossplane-resources        | 0.1.0               | Unchanged           |
| cilium-servicemonitors             | 0.1.2               | Unchanged           |
| cloud-provider-aws                 | 1.29.3-gs1          | 1.30.3-gs1          |
| cluster-autoscaler                 | 1.29.3-gs1          | 1.30.3-gs1          |
| coredns                            | 1.22.0              | 1.23.0              |
| etcd-k8s-res-count-exporter        | 1.10.0              | Unchanged           |
| external-dns                       | 3.1.0               | Unchanged           |
| irsa-servicemonitors               | 0.1.0               | Unchanged           |
| k8s-audit-metrics                  | 0.10.0              | Unchanged           |
| k8s-dns-node-cache                 | 2.8.1               | Unchanged           |
| metrics-server                     | 2.4.2               | Unchanged           |
| net-exporter                       | 1.21.0              | Unchanged           |
| network-policies                   | 0.1.1               | Unchanged           |
| node-exporter                      | 1.20.0              | Unchanged           |
| observability-bundle               | 1.8.0               | 1.9.0               |
| observability-policies             | 0.0.1               | Unchanged           |
| prometheus-blackbox-exporter       | 0.4.2               | Unchanged           |
| security-bundle                    | 1.8.2               | 1.9.1               |
| teleport-kube-agent                | 0.10.3              | Unchanged           |
| vertical-pod-autoscaler            | 5.3.0               | Unchanged           |
| vertical-pod-autoscaler-crd        | 3.1.1               | Unchanged           |
+------------------------------------+---------------------+---------------------+
+----------------+-----------------+----------------------------+
| COMPONENT NAME | CURRENT VERSION | DESIRED VERSION            |
+----------------+-----------------+----------------------------+
| cluster-aws    | 2.4.0           | Unchanged                  |
| flatcar        | 3975.2.2        | 4081.2.0                   |
| kubernetes     | 1.29.10         | 1.30.7 - requested by user |
| os-tooling     | 1.20.1          | 1.21.1                     |
+----------------+-----------------+----------------------------+
Do you want to continue? (y/n)y
```

**Creates**

```
capa/v29.5.0/
|-- README.md
|-- announcement.md
|-- kustomization.yaml
|-- release.diff
`-- release.yaml
```

**Modifies**

```
capa/kustomization.yaml
capa/releases.json # will be sorted when creating older releases
```

**Preview releases.json**

```
...
    {
      "version": "29.5.0",
      "isDeprecated": false,
      "releaseTimestamp": "2024-12-10T15:53:41+01:00",
      "changelogUrl": "https://github.com/giantswarm/releases/blob/master/aws/v29.5.0/README.md",
      "isStable": true
    }
...
```

**Preview announcement.md**

```
**Workload cluster release v29.5.0 for CAPA is available**. << Add description here >> 

Further details can be found in the [release notes](https://docs.giantswarm.io/changes/workload-cluster-releases-capa/releases/aws-29.5.0).
```

**Preview README.md**

image

I separated components, apps and added a short list of the apps and components which have changed before mentioning the changes of each component and app

### Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Review by anvddriesch — APPROVED** [review]

File: pkg/release/create.go
Comment (AndiDog) [inline pkg/release/create.go:26]: I guess this means that creating vintage releases won't work anymore. But we can still revert to old binaries to do that, or handcraft them, so it's fine IMO.

File: pkg/release/create.go
Comment (AndiDog) [inline pkg/release/create.go:193]: The `releases.json` file can currently be regenerated in `cd releases && go run renovate.go` AFAIK. Should we throw away one of the implementations?

(Not blocking this PR.)

**Review by AndiDog — APPROVED** [review]

File: pkg/release/create.go
Comment (njuettner) [reply pkg/release/create.go:26]: Yeah, my thinking was we don't need another vintage release. Using an old binary was also my idea, I make it explicit in the changelog to use an older release of devctl if we need a new vintage release

File: pkg/release/create.go
Comment (njuettner) [reply pkg/release/create.go:193]: IMHO we can drop `go run renovate.go` after the we merge this PR

**Review by AndiDog — COMMENTED** [review]
If I "accidentally" use an unexpected provider (`devctl release create --provider capa`), I get no error, but unfilled templates like `# :zap: Giant Swarm Release v25.1.3 for :zap:`. We should add parameter validation.

**Comment by AndiDog** [comment]
Also, if I add a new application such as `--app aws-nth-bundle@1.2.0` which the previous release didn't have, I get this in the announcement:

```
* aws-nth-bundle from  to [1.2.0](https://github.com/giantswarm/aws-nth-bundle/releases/tag/v1.2.0)
```

but should say "Added [...]".

**Comment by njuettner** [comment]
@AndiDog good input, thanks. I see you added `aws-nth-bundle`. 

I changed the template slightly in case something new is added it will be specified as
`Added aws-nth-bundle [1.2.0](https://github.com/giantswarm/aws-nth-bundle/releases/tag/v1.2.0)`

I also added a check to verify the provider which is specified should match one of the supported ones.

**Comment by njuettner** [comment]
I also changed the `kustomization.yaml` file so it matches with our latest changes

**Comment by Gacko** [comment]
Yeah, I guess this is the best approach. Because otherwise nobody will use it and we probably don't get a perfect solution on the first try. 🙂

**Pull request merged**
